gprofiler/metadata/application_metadata.py

```diff
@@ -0,0 +1,52 @@
+#
+# Copyright (c) Granulate. All rights reserved.
+# Licensed under the AGPL3 License. See LICENSE.md in the project root for license information.
+#
+
+from threading import Event, Lock
+from typing import Any, Dict, Optional, Union
+
+from granulate_utils.linux.process import is_process_running
+from psutil import NoSuchProcess, Process
+
+from gprofiler.utils.elf import get_process_execfn
+
+
+def get_application_metadata(process: Union[int, Process]) -> Optional[Dict]:
+    if process in (0, -1):  # funny values retrieved by perf
+        return None
+
+    try:
+        process = process if isinstance(process, Process) else Process(process)
+    except NoSuchProcess:
+        return None
+
+    return ApplicationMetadata.get_metadata(process)
+
+
+class ApplicationMetadata:
+    _CACHE_CLEAR_ON_SIZE = 16384
+    _cache: Dict[Process, Optional[Dict]] = {}
+    _cache_clear_lock = Lock()
+
+    @classmethod
+    def get_metadata(cls, process: Process) -> Optional[Dict]:
+        return cls._cache.get(process)
```

Comment:
Why not to acquire the lock here?

Reply:
Why yes? It either works, or doesn't work. No race here

Reply:
No risk of data corruption? I don't know what's considered "atomic" in Python, but I can imagine possible data corruption if `get`ting while updating for example.
What's the race you're preventing with the clear cache lock? Iterator invalidation in the `for` loop?